certbot-ci/certbot_integration_tests/certbot_tests/test_main.py

```diff
@@ -19,21 +23,63 @@ def context(request):
         integration_test_context.cleanup()
 
 
-def test_manual_http_auth(context):
-    """Test the HTTP-01 challenge using manual plugin."""
-    with misc.create_http_server(context.http_01_port) as webroot:
-        manual_http_hooks = misc.manual_http_hooks(webroot)
+def test_manual_dns_auth(context):
+    """Test the DNS-01 challenge using manual plugin."""
+    certname = context.domain('dns')
+    context.certbot([
+        '-a', 'manual', '-d', certname, '--preferred-challenges', 'dns',
+        'run', '--cert-name', certname,
+        '--manual-auth-hook', context.manual_dns_auth_hook,
+        '--manual-cleanup-hook', context.manual_dns_cleanup_hook,
+        '--pre-hook', 'echo wtf.pre >> "{0}"'.format(context.hook_probe),
+        '--post-hook', 'echo wtf.post >> "{0}"'.format(context.hook_probe),
+        '--renew-hook', 'echo renew >> "{0}"'.format(context.hook_probe)
+    ])
 
-        certname = context.domain()
-        context.certbot([
-            'certonly', '-a', 'manual', '-d', certname,
-            '--cert-name', certname,
-            '--manual-auth-hook', manual_http_hooks[0],
-            '--manual-cleanup-hook', manual_http_hooks[1],
-            '--pre-hook', 'echo wtf.pre >> "{0}"'.format(context.hook_probe),
-            '--post-hook', 'echo wtf.post >> "{0}"'.format(context.hook_probe),
-            '--deploy-hook', 'echo deploy >> "{0}"'.format(context.hook_probe)
-        ])
+    with pytest.raises(AssertionError):
+        assert_hook_execution(context.hook_probe, 'renew')
+    assert_save_renew_hook(context.config_dir, certname)
+
+
+def test_renew(context):
+    """Test various certificate renew scenarios."""
```

Comment:
Looks like there's a few different tests in here. Any reason these can't be different functions, with setup/checks abstracted?

Reply:
The original test is even longer than that. No reason to keep this big chunk of code together at the end. Except that the more I will refactor here, the more it will be away from the original bash script. 

I would like first to have the code functional be merged, and refactor in a future PR.

Reply:
I'd like to push back against the idea of committing code now that we plan to refactor later. I think it's possible to end up with clean, final-version-worthy code that still covers the original things that are tested, in a way that lets us implement it incrementally. I'd rather see only one small piece of it in a single function, along with the setup/cleanup code in the places it'll eventually be, than see all of the code dumped in at once and its framework abstracted out later.

Reply:
See below.